companion/src/modeledit/expodialog.cpp

```diff
@@ -140,71 +144,84 @@ void ExpoDialog::updateScale()
-  ed->swtch = RawSwitch(ui->switchesCB->itemData(ui->switchesCB->currentIndex()).toInt());
-  ed->mode = ui->sideCB->currentIndex() + 1;
+  if (!lock) {
+    lock = true;
+    QCheckBox * cb_fp[] = {ui->cb_FP0,ui->cb_FP1,ui->cb_FP2,ui->cb_FP3,ui->cb_FP4,ui->cb_FP5,ui->cb_FP6,ui->cb_FP7,ui->cb_FP8 };
+
+    RawSource srcRaw = RawSource(ui->sourceCB->itemData(ui->sourceCB->currentIndex()).toInt());
+    if (ed->srcRaw != srcRaw) {
+      ed->srcRaw = srcRaw;
+      updateScale();
+    }
 
-  strcpy(ed->name, ui->lineName->text().toAscii().data());
-  if (firmware->getCapability(VirtualInputs)) {
-    inputName = ui->inputName->text();
-  }
+    RawSourceRange range = srcRaw.getRange(&model, generalSettings);
+    ed->scale = round(float(ui->scale->value()) / range.step);
+    ed->carryTrim = 1 - ui->trimCB->currentIndex();
+    ed->swtch = RawSwitch(ui->switchesCB->itemData(ui->switchesCB->currentIndex()).toInt());
+    ed->mode = ui->sideCB->currentIndex() + 1;
 
-  ed->flightModes = 0;
-  for (int i=8; i>=0 ; i--) {
-    if (!cb_fp[i]->checkState()) {
-      ed->flightModes++;
-    }
-    ed->flightModes <<= 1;
-  }
-  ed->flightModes >>= 1;
-  if (firmware->getCapability(FlightModes)) {
-    int zeros = 0;
-    int ones = 0;
-    int phtemp = ed->flightModes;
-    for (int i=0; i<firmware->getCapability(FlightModes); i++) {
-      if (phtemp & 1) {
-        ones++;
-      }
-      else {
-        zeros++;
-      }
-      phtemp >>= 1;
+    strcpy(ed->name, ui->lineName->text().toAscii().data());
+    if (firmware->getCapability(VirtualInputs)) {
+      inputName = ui->inputName->text();
     }
-    if (zeros == 1) {
-      phtemp=ed->flightModes;
-      for (int i=0; i<firmware->getCapability(FlightModes); i++) {
-        if ((phtemp & 1) == 0) {
-          break;
-        }
-        phtemp >>= 1;
-      }
-    }
-    else if (ones == 1) {
-      phtemp = ed->flightModes;
-      for (int i=0; i<firmware->getCapability(FlightModes); i++) {
-        if (phtemp & 1) {
-          break;
-        }
-        phtemp >>= 1;
+
+    ed->flightModes = 0;
+    for (int i=8; i>=0 ; i--) {
+      if (!cb_fp[i]->checkState()) {
+        ed->flightModes++;
       }
+      ed->flightModes <<= 1;
     }
+    ed->flightModes >>= 1;
+    lock = false;
   }
-  else {
-    ed->flightModes = 0;
-  }  
 }
 
 void ExpoDialog::shrink()
 {
   resize(0, 0);
 }
+void ExpoDialog::label_phases_customContextMenuRequested(const QPoint & pos)
+{
+  QLabel *label = (QLabel *)sender();
+  QPoint globalPos = label->mapToGlobal(pos);
+  QMenu contextMenu;
+  contextMenu.addAction(tr("Clear All"), this, SLOT(fmClearAll()));
+  contextMenu.addAction(tr("Set All"), this, SLOT(fmSetAll()));
+  contextMenu.addAction(tr("Invert All"), this, SLOT(fmInvertAll()));
+  contextMenu.exec(globalPos);
+}
+
+void ExpoDialog::fmClearAll()
+{
+  lock = true;
+  QCheckBox * cb_fp[] = {ui->cb_FP0,ui->cb_FP1,ui->cb_FP2,ui->cb_FP3,ui->cb_FP4,ui->cb_FP5,ui->cb_FP6,ui->cb_FP7,ui->cb_FP8 };
```

Comment:
perhaps you could make a method which would return this array? in order to avoid duplicated code